pkg/client/namespaced_client.go

```diff
@@ -0,0 +1,156 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package client
+
+import (
+	"context"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+// NewNamespacedClient wraps an existing client enforcing the namespace value.
+// All functions using this client will have the same namespace declared here.
+func NewNamespacedClient(c Client, ns string) Client {
+	return &namespacedClient{
+		client:    c,
+		namespace: ns,
+	}
+}
+
+var _ Client = &namespacedClient{}
+
+// namespacedClient is a Client that wraps another Client in order to enforce the specified namespace value.
+type namespacedClient struct {
+	namespace string
+	client    Client
+}
+
+// Create implements clinet.Client
+func (n *namespacedClient) Create(ctx context.Context, obj runtime.Object, opts ...CreateOption) error {
+	metaObj, err := meta.Accessor(obj)
+	if err != nil {
+		return err
+	}
+
+	if n.namespace != "" && metaObj.GetNamespace() != n.namespace {
+		metaObj.SetNamespace(n.namespace)
+	}
+	return n.client.Create(ctx, obj, opts...)
+}
+
+// Update implements client.Client
+func (n *namespacedClient) Update(ctx context.Context, obj runtime.Object, opts ...UpdateOption) error {
+	metaObj, err := meta.Accessor(obj)
+	if err != nil {
+		return err
+	}
+
+	if n.namespace != "" && metaObj.GetNamespace() != n.namespace {
```

Comment:
I think rather than trying to infer from the object if its namespaced by checking if any namespace is set, it would be more logical to take a [`RESTMapper`](pkg/client/apiutil/dynamicrestmapper_test.go) in the constructor of this wrapper and use its `RESTMapping` method to figure out if a given object is namespaced or not (You can use [`GVKForObject`](https://github.com/alvaroaleman/controller-runtime/blob/accbce34c7c0a807ad2beca5e7cf9672aa914a18/pkg/client/apiutil/apimachinery.go#L50) to get the gvk)

Reply:
Thanks for the review @alvaroaleman. Had a doubt regarding this. `GVKForObject` requires `runtime.Scheme`, which is useful in `RestMapping` method too. Currently we don't have a method to obtain the `scheme` from client Options, when it is [generated](https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/client?tab=doc#New). If we go ahead with creating a new scheme in the implementation, we would have to register the object Kind manually using [`AddToScheme`](https://pkg.go.dev/k8s.io/apimachinery/pkg/runtime?tab=doc#SchemeBuilder.AddToScheme). Is there any another way to go about this ?

Reply:
Related #1058

Reply:
I think for now the `constructor` needs to take both a `RESTMapper` and a `Scheme`. Ideally we could just obtain both of those from the Client, but that is not the case today. Once #1058 is in, we can at least obtain the Scheme from the Client.

Reply:
Thanks @vincepri and @alvaroaleman . Will update the PR.

Reply:
I see that in the implementation of [`client.New(..)`](https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/client?tab=doc#New), `runtime.Scheme` will not be used for unstructured objects. Hence, even when we pass a `scheme` with the `client`, GVK is not registered - similar to test case present [here](https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/client/client_test.go#L525). If this is the case, then wouldn't `RESTMapping` not be useful for unstructured objects, since for them GVK is extracted from Object itself ?

Reply:
With unstructured objects, the user is required to set the GVK before passing in to a client, so you should be able to use `GetObjectKind()` on those.

Reply:
That said, exposing a RESTMapper from a Client is probably a good thing. We can follow the approach used in #1058 to work out something similar

Reply:
We've merged https://github.com/kubernetes-sigs/controller-runtime/pull/1109/files, which brings in support to get a RESTMapper from a client